cpp/include/raft/comms/std_comms.hpp

```diff
@@ -0,0 +1,406 @@
+  }
+
+  void bcast(void *buff, int count, datatype_t datatype, int root,
+             cudaStream_t stream) const {
+    NCCL_CHECK(ncclBroadcast(buff, buff, count, getNCCLDatatype(datatype), root,
+                             _nccl_comm, stream));
+  }
+
+  void reduce(const void *sendbuff, void *recvbuff, int count,
+              datatype_t datatype, op_t op, int root,
+              cudaStream_t stream) const {
+    NCCL_CHECK(ncclReduce(sendbuff, recvbuff, count, getNCCLDatatype(datatype),
+                          getNCCLOp(op), root, _nccl_comm, stream));
+  }
+
+  void allgather(const void *sendbuff, void *recvbuff, int sendcount,
+                 datatype_t datatype, cudaStream_t stream) const {
+    NCCL_CHECK(ncclAllGather(sendbuff, recvbuff, sendcount,
+                             getNCCLDatatype(datatype), _nccl_comm, stream));
+  }
+
+  void allgatherv(const void *sendbuf, void *recvbuf, const int recvcounts[],
+                  const int displs[], datatype_t datatype,
+                  cudaStream_t stream) const {
+    //From: "An Empirical Evaluation of Allgatherv on Multi-GPU Systems" - https://arxiv.org/pdf/1812.05964.pdf
+    //Listing 1 on page 4.
+    for (int root = 0; root < _size; ++root) {
+      size_t dtype_size = getDatatypeSize(datatype);
+      NCCL_CHECK(ncclBroadcast(
+        sendbuf, static_cast<char *>(recvbuf) + displs[root] * dtype_size,
+        recvcounts[root], getNCCLDatatype(datatype), root, _nccl_comm, stream));
+    }
+  }
+
+  void reducescatter(const void *sendbuff, void *recvbuff, int recvcount,
+                     datatype_t datatype, op_t op, cudaStream_t stream) const {
+    NCCL_CHECK(ncclReduceScatter(sendbuff, recvbuff, recvcount,
+                                 getNCCLDatatype(datatype), getNCCLOp(op),
+                                 _nccl_comm, stream));
+  }
+
+  status_t syncStream(cudaStream_t stream) const {
+    cudaError_t cudaErr;
+    ncclResult_t ncclErr, ncclAsyncErr;
+    while (1) {
+      cudaErr = cudaStreamQuery(stream);
+      if (cudaErr == cudaSuccess) return status_t::commStatusSuccess;
+
+      if (cudaErr != cudaErrorNotReady) {
+        // An error occurred querying the status of the stream
+        return status_t::commStatusError;
+      }
+
+      ncclErr = ncclCommGetAsyncError(_nccl_comm, &ncclAsyncErr);
+      if (ncclErr != ncclSuccess) {
+        // An error occurred retrieving the asynchronous error
+        return status_t::commStatusError;
+      }
+
+      if (ncclAsyncErr != ncclSuccess) {
+        // An asynchronous error happened. Stop the operation and destroy
+        // the communicator
+        ncclErr = ncclCommAbort(_nccl_comm);
+        if (ncclErr != ncclSuccess)
+          // Caller may abort with an exception or try to re-create a new communicator.
+          return status_t::commStatusAbort;
+      }
+
+      // Let other threads (including NCCL threads) use the CPU.
+      pthread_yield();
```

Comment:
Better use standard C++ function (https://en.cppreference.com/w/cpp/thread/yield) than pthread?

Reply:
This was taken directly from the NCCL docs

Reply:
I see, but I think it is better to go for language standard than a third-party (not sure we can call pthread third-party but anyways) library.